src/amd/compiler/aco_util.h

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright Michael Schellenberger Costa
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice (including the next
+ * paragraph) shall be included in all copies or substantial portions of the
+ * Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ *
+ */
+
+#ifndef ACO_UTIL_H
+#define ACO_UTIL_H
+
+#include <utility>
+
+namespace aco {
+
+/*!	\brief		Definition of a span object
+*
+* \details		A "span" is an "array view" type for holding a view of a contiguous data
+*				array. the "span" object does not and cannot own such data itself. A span
+*				provides an easy way to provide overloads for anything operating on
+*				contiguous sequences without needing to manage pointers and array lengths
+*				manually.
+
+*				A span is conceptually a pointer (ptr) and a length (size) into an already
+*				existing array of contiguous memory; the array it represents references the
+*				elements "ptr[0] .. ptr[size-1]". Passing a properly-constructed "span"
+*				instead of raw pointers avoids many issues related to index out of bounds
+*				errors.
+*/
+template <typename T>
+class span {
+public:
+	using value_type             = T;
+	using pointer                = value_type*;
+	using const_pointer          = const value_type*;
+	using reference              = value_type&;
+	using const_reference        = const value_type&;
+	using iterator				 = pointer;
```

Comment:
Is this file supposed to use tabs while all other files use spaces?

Reply:
It should probably use three spaces for indentation instead of tabs

Reply:
Yeah, I wrote it for a personal project where i use tabs. I will fix that

Reply:
Do I see right that this whole thing is basically re-implementing std::span from c++20?

Reply:
Yes and no. This is a subset of `std::span`, which ommits the additional template parameter that represents the extend of the multidimensional `std::span`

Reply:
I replaced tabs with 3 spaces and shortened the details section